cognite/extractorutils/rest/extractor.py

```diff
@@ -0,0 +1,160 @@
+from cognite.extractorutils.configtools import BaseConfig
+from cognite.extractorutils.uploader import EventUploadQueue
+from more_itertools import peekable
+
+from cognite.extractorutils import Extractor
+from cognite.extractorutils.rest.http import HttpCall, HttpMethod, HttpUrl, RequestBody, ResponseType
+
+FullFile = Union[Tuple[FileMetadata, PathLike], Tuple[FileMetadata, bytes]]
+
+CdfTypes = Union[
+    Event, Iterable[Event], Datapoint, Iterable[Datapoint], Row, Iterable[Row], FullFile, Iterable[FullFile]
+]
+
+
+@dataclass
+class Endpoint(Generic[ResponseType]):
+    implementation: Callable[[ResponseType], CdfTypes]
+    method: HttpMethod
+    path: str
+    query: Dict[str, Any]
+    body: Optional[RequestBody]
+    response_type: Type[ResponseType]
+    next_url: Optional[Callable[[str, ResponseType], Optional[str]]]
+    interval: Optional[int]
+
+
+@dataclass
+class SourceConfig:
+    # TODO: Actually use these, and add them to request headers
+    idp_authentication: Optional[AuthenticatorConfig] = None
+    headers: Optional[Dict[str, Any]] = None
+
+
+@dataclass
+class RestConfig(BaseConfig):
+    source: SourceConfig = SourceConfig()
+
+
+class RestExtractor(Extractor):
+    def __init__(
+        self,
+        *,
+        name: str,
+        description: str,
+        version: Optional[str] = None,
+        cancelation_token: Event = threading.Event(),
+        base_url: Optional[str],
+        headers: Optional[Dict[str, Union[str, Callable[[], str]]]] = None,
+    ):
+        super(RestExtractor, self).__init__(
+            name=name,
+            description=description,
+            version=version,
+            cancelation_token=cancelation_token,
+            use_default_state_store=False,
+            config_class=RestConfig,
+        )
+        self.base_url = base_url or ""
+        self.headers = headers or {}
+        self.endpoints: List[Endpoint] = []
+
+    def endpoint(
+        self,
+        *,
+        method: HttpMethod,
+        path: str,
+        query: Dict[str, Any],
+        body: Optional[RequestBody],
+        response_type: Type[ResponseType],
+        next_url: Optional[Callable[[str, ResponseType], Optional[str]]],
+        interval: Optional[int],
+    ):
+        def decorator(func):
+            self.endpoints.append(
+                Endpoint(
+                    implementation=func,
+                    method=method,
+                    path=path,
+                    query=query,
+                    body=body,
+                    response_type=response_type,
+                    next_url=next_url,
+                    interval=interval,
+                )
+            )
+            return func
+
+        return decorator
+
+    def get(
+        self,
+        path: str,
+        *,
+        query: Optional[Dict[str, Any]] = None,
+        response_type: Type[ResponseType],
+        next_url: Optional[Callable[[HttpCall], Optional[HttpUrl]]] = None,
```

Comment:
There should be some solution for the initial url, beyond `query`, since it can change between runs. Either let `path` be callable, and remove `query` and `next_url`, like you mentioned, or create an `initial_url` method.

Reply:
Agreed! 

I'm inclined to let `path` itself be a callable. In fact, I'm thinking about ways were we could allow (at least almost) every parameter to be either a value or a callable yielding a value. That would be awesome for keeping it generic enough to work in most circumstances.

Reply:
so let e.g. the `path` arg be `Union[str, Callable[[HttpCall], str]` and `headers` arg be `Dict[str, Union[str, Callable[[], str]]]` and so on

Reply:
it might lead to some not-so-nice branching within the library, which is what I'm trying to figure out how to do nice at the moment, but it would be really cool from a user perspective IMO

Reply:
Agreed. For the library you could have a helper method `from_optional_callable` or something like that, which takes an `Union[type, Callable[[SomeInput], type]`, maybe? Just to simplify the code.